File: Javascript/0977_Squares _of_a_Sorted_Array.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var sortedSquares = function(nums) {
    let l = 0;
    let r = nums.length - 1;
    let sol = [];
    while(l <= r) {
        if(nums[l]**2 >= nums[r]**2) {
            sol.unshift(nums[l]**2);
            l++;
        }
        else {
            sol.unshift(nums[r]**2);
            r--;
        }
    }
    return sol;
};

File: Python/0977_Squares_of_a_Sorted_Array.py
class Solution(object):
    def sortedSquares(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        # Time Complexity: O(N)
        # Space Complexity: O(N)
        n = len(nums)
        ans = [0] * n

        left, right = 0, n-1
        
        for i in range(n-1, -1, -1):
            if abs(nums[left]) >= abs(nums[right]):
                ans[i] = nums[left] ** 2
                left += 1
            else:
                ans[i] = nums[right] ** 2
                right -= 1

        return ans